robot/board.py

```diff
@@ -46,43 +47,53 @@ def _send(self, message, retry=False):
         try:
             self.sock.send(message)
         except (socket.timeout, BrokenPipeError, OSError):
-            if retry:
+            if _is_retry:
                 raise ConnectionError(self._get_lc_error())
             else:
                 try:
                     self._connect(self.sock_path)  # Reconnect
                 except FileNotFoundError:
                     raise ConnectionError(self._get_lc_error())
-                self._send(message, retry=True)  # Retry Recursively
+                self._send(message, _is_retry=True)  # Retry Recursively
 
-    def _recv(self, _retry=False):
-        """
-        Receive a message from the robotd socket
-        :return: the message
-        """
+    def _socket_with_single_retry(self, handler):
+        retryable_errors = (socket.timeout, BrokenPipeError, OSError)
+
+        try:
+            return handler(self.sock)
+        except retryable_errors:
+            pass
 
-        # TODO catch empty string receives as bad things
-        unended_text = b''
+        # Retry once, need to reconnect first
+        try:
+            self._connect(self.sock_path)
+        except FileNotFoundError:
+            raise ConnectionError(self._get_lc_error())
 
-        while not self._recv_buffer:
-            strings = []
-            try:
-                output = self.sock.recv(Board.RECV_BUFFER)
-            except (socket.timeout, BrokenPipeError, OSError) as e:
-                if _retry:
-                    raise ConnectionError(self._get_lc_error())
-                else:
-                    print("Connection", self.sock_path, e, "retrying")
-                    self._connect(self.sock_path)
-                    return self._recv(_retry=True)  # Retry recursively
-            strings.extend((unended_text + output).split(b'\n'))
-            if strings == [b'']:
-                self._recv_buffer.append(b'')
-            unended_text = strings.pop()
-            if not unended_text:
-                self._recv_buffer.extend([x for x in strings if x != b''])
-
-        return self._recv_buffer.pop(0)
+        try:
+            return handler(self.sock)
+        except retryable_errors:
+            raise ConnectionError(self._get_lc_error())
+
+    def receive_raw_from_socket_with_single_retry(self):
+        return self._socket_with_single_retry(
+            lambda s: s.recv(Board.RECV_BUFFER_BYTES),
+        )
+
+    def _recv(self, should_retry=True):
+        while b'\n' not in self.data:
+            if should_retry:
+                message = self.receive_raw_from_socket_with_single_retry()
+            else:
+                message = self.sock.recv()
+            if message is b'':
+                # Received blank, return blank
+                return b''
+
+            self.data += message
+        line = re.search(b'.*\n', self.data).group(0)
```

Comment:
You might enjoy the `maxsplit` argument to [`split`](https://docs.python.org/3/library/stdtypes.html#str.split) which would simplify this to:
``` python
line, self.data = self.data.split(b'\n', 1)
```